### PR TITLE
MINOR: Throw ProducerFencedException directly but with a new instance

### PR DESCRIPTION
**Pull request by guozhangwang**
Currently in maybeFailWithError, we always wrap the lastError as a KafkaException. For ProducerFencedException however, we should just throw the exception itself; however we throw a new instance since the previous book-kept one's call trace is not from this call, and hence could be confusing.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@hachikuji @abbccdda for reviews.

**Review by abbccdda — APPROVED** [review]
Thank you for the patch, LGTM

**Comment by abbccdda** [comment]
The test failure seems irrelevant. Let me trigger a retest.

**Comment by abbccdda** [comment]
Retest this please

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 1]
```diff
@@ -1378,7 +1379,44 @@ public void testProducerFencedException() throws InterruptedException, Execution
...
+            responseFuture.get();
+            fail("Expected to get a ExecutionException from the response");
+        } catch (ExecutionException e) {
+            assertTrue(e.getCause() instanceof ProducerFencedException);
+        }
+
+        // make sure the produce was expired.
+        try {
+            transactionManager.beginTransaction();
+            fail("Expected to get a ExecutionException from the beginTransaction call");
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: We can use `assertThrows`. It is much more concise.

**Review by hachikuji — APPROVED** [review]
Just a minor comment. LGTM overall.

**Pull request merged**